### PR TITLE
fix(hyperlink): fix hyperlinks on Windows

### LINKED ISSUE #335: bug: --hyperlink links all raise an "invalid link" error when clicked
- eza --version = 0.12.0
- The command-line arguments you are using: --hyperlink
- Your operating system and hardware platform: Windows 11 Pro + Windows Terminal 1.17.11461.0 on HP ZBook Firefly 15 G8 PC

Hello,
any hyperlink generated leads to this same "the link is not valid" error when clicked - be it a directory or a file of any type:

image

### PR DESCRIPTION
**Pull request by daviessm**
Fixes #335

### CODE REVIEW

**Review by PThorpe92 — APPROVED** [review]
Nice! thx

**Comment by ariasuni** [comment]
I would have liked being added as a reviewer on this one, since I last modified this part of the code (and that I will need to make a fix PR, because the file names still need to be escaped, and the hostname is still a recommended part of the `file://` URL for hyperlinks).

However, my bad for breaking Windows.

File: src/output/file_name.rs [hunk 1]
```diff
@@ -351,13 +351,22 @@ impl<'a, 'dir, C: Colours> FileName<'a, 'dir, C> {
...
                     HYPERLINK_START,
                     gethostname::gethostname().to_str().unwrap_or(""),
                     urlencoding::encode(abs_path).replace("%2F", "/"),
                     HYPERLINK_END,
                 )));
+                #[cfg(target_os = "windows")]
+                bits.insert(0, ANSIString::from(format!(
+                    "{}file://{}{}",
+                    HYPERLINK_START,
+                    abs_path.replace("\\\\?\\", ""),
```
Comment (ariasuni) [inline src/output/file_name.rs:366]: What does the `.replace()` part do? It’s unclear to me. @daviessm

Comment (daviessm) [reply src/output/file_name.rs:366]: `Path::canonicalize` is an alias for `fs::canonicalize` and on Windows that returns a path that begins with `\\?\`, e.g. `\\?\C:\Users`. I couldn't find any documentation on how or whether that works with `file://` URLs and from my testing it doesn't, so I just stripped it off. This means that very long paths will probably not work, but they break things in Windows anyway so it isn't really a big issue.

**Comment by daviessm** [comment]
> I would have liked being added as a reviewer on this one, since I last modified this part of the code (and that I will need to make a fix PR, because the file names still need to be escaped, and the hostname is still a recommended part of the `file://` URL for hyperlinks).
> 
> However, my bad for breaking Windows.

When I was testing this the only way I could get it to work was the above. It seems to work with spaces and everything; when there was a hostname or the path was urlescaped the hyperlinks didn't work at all. I couldn't find any documentation on how `file://` URLs are supposed to work on Windows though.

**Comment by ariasuni** [comment]
Well it doesn’t surprise me that it works without url encoding for simple filenames, but I’m worried it might not work for more complex ones. Indeed, the `file://` protocol doesn’t seem to support `\\?\.

Actually, I think the code could work by simply using `abs_path.replace("\\\\?\\", "").replace("%2F", "/")` and I’ll test this hypothesis a bit later, for now it should work on most files anyway (as per your testing).

[Source](https://en.wikipedia.org/wiki/File_URI_scheme#Windows_2)

**Comment by RuRo** [comment]
@ariasuni The `file://hostname/path` thing isn't guaranteed to work on linux either. I think, most users have their hostname configured as the name of their device, but AFAIK it's pretty rare to have that name in `/etc/hosts` or otherwise available via DNS. So for a lot of users, the hyperlinks will attempt to open something like `file://my_laptop_name/home/users/somefile.txt` and it will fail.

I think adding `hostname` to the link should be optional, and it should use `file:///` (three slashes) by default.

**Comment by ariasuni** [comment]
Well, I was already thinking about removing hostname entirely, and you’re adding a convincing argument on top of that…

**Pull request merged**